### ISSUE TITLE
Advice or help with creating an HCL parser with variables and dynamic blocks

### ISSUE THREAD

**Issue by jlarfors** [hashicorp/hcl#401]
We are developing an open source project that will use some more advanced features of HCL, and could really do with some help from someone more experienced to double-check what we are doing.

Our use case is that we want to an HCL parser that supports modules (like Terraform modules) and variables & functions (`EvalContext` stuff).

We have played around quite a lot with different methods of decoding HCL, e.g. into a Go struct, into cty.Value (using `hcldec` specs), and manually traversing HCL using a schema (such as `hcl.BodySchema`).

Also spending a lot of time digging through the Terraform code base has taught us quite a lot and we would like to create a nice approach for this, that could possibly be reused by others in the future (e.g. we do not need Terraform's support for external providers).

## Background

Let's take a really simple example where the only blocks are `input` and `output`. Let's define the go structs below:

```go
type Input struct { /* ... */ }
type Output struct { /* ... */ }

// a module should be processed into an instance of this
type HCLFile struct {
	Inputs   []Input   `hcl:"input,block"`
	Outputs []Output `hcl:"output,block"`
}
```

And take an example hcl file:

```hcl
input "ex_input" {}

output "ex_output" {
  value = someoperation(input.ex_input)
}
```

To decode the HCL into an instance of `HCLFile` (which is our goal) we need to create an `EvalContext` which includes the variable `input.ex_input` and the function `someoperation(...)`.

The easiest approach we have found to traverse HCL is to create a schema for it (e.g. `hcl.BodySchema`) and then iterate over the `Blocks` and `Attributes` recursively.

To create the `EvalContext` we need to create `cty.Value` for the variables for which we could use `hcldec.Spec` to decode HCL bodies. The functions in `EvalContext` will be defined beforehand, so that's easy.

## Proposed Solution

The idea we came up with is the use Go reflection on the struct `HCLFile` and from this be able to create an HCL schema and `hcldec.Spec` for the different bodies.

Then traverse the HCL blocks and attributes recursively using the schema, and create a minimal `EvalContext` by converting each body into a `cty.Value` (ignoring the HCL expressions that reference variables - found using `Expr.Variables()` to get `[]Traversal`).

Then have some kind of dependency graph between variable references in HCL so that we can resolve variables (and therefore HCL bodies) using the initial `EvalContext`, and add these new variables to the existing `EvalContext` - the goal being to create an `EvalContext` that contains everything.

Finally at the end we just `gohcl.DecodeBody(...)` and decode the whole thing into `HCLFile` using our created `EvalContext`

## Considerations

We believe that Terraform uses it's own `Block` (and other) structs which are used to dynamically create the HCL schema and parse things using that, but we do not have the need for extensions through providers. We just want the easiest way to use this, whilst allowing functions and variables. So we haven't wanted to copy Terraform directly.

Is there any reason why this could be a limiting or bad approach? Is there a better way that someone could suggest?
Has someone already done something similar?

Any help or feedback appreciated and we would love to get some more advanced examples or wrappers for HCL written to make this easier for others in the future!

This is quite a lot for me to write, and for readers to take in, so I am happy to clarify anything in the above.

**Comment by jlarfors** [comment]
This issue can be closed as we have since solved this. We have taken inspiration from the `gohcl` package and created some of our adjustments.

It might be useful if some of the methods like `getFieldTags(reflect.Type)` could be exported, for which I will create another issue.

**Comment by vivekskrishna** [comment]
Hi @jlarfors . have you open sourced your work or is it a private implementation. if it is open sourced can you share the repo.we are looking for a similar library to do some parsing.

**Comment by jlarfors** [comment]
@vivekskrishna it is open source, but the project ended and by the end we avoided HCL, and I would encourage you to reconsider HCL for anything dynamic.
In the end we went with Rego (from OPA).

What terraform does is a whole lot on top of HCL, and we spent a lot of time trying to do similar things, and that was time that could've spent spent on much more useful things.

So take it from someone who was a fan of HCL, burnt his fingers, and now tried to avoid it for building anything more than a simple config language.
